agent/util-scripts/pbench-tool-meister-start

```diff
@@ -527,150 +616,333 @@ def main(argv):
-        ret_val = kill_redis_server(redis_pid)
-    elif successes > 0:
-        # If any successes, then we need to wait for them to show up as
-        # subscribers.
-        logger.debug(
-            "e. waiting for all successfully spawned SSH processes"
-            " to show up as subscribers"
+            return kill_redis_server(redis_pid, ReturnCode.TMNOSUCCESSES)
+
+        assert successes == tm_count, (
+            f"Logic Bomb! Number of created Tool Meisters, {successes}, does not"
+            f" match the expected number of Tool Meisters, {tm_count}"
         )
-        pids = wait_for_subs(chan, successes, logger)
-        # Record our collected pids.
-        try:
-            redis_server.set("tm-pids", json.dumps(pids, sort_keys=True))
-        except Exception:
-            logger.exception("failed to set tool meister pids object")
-            ret_val = kill_redis_server(redis_pid)
+
+    # +
+    # 6. Wait for the TDS to send a message reporting that it, and all the
+    # TMs, started.
+    # -
+
+    # If any successes, then we need to wait for them to show up as
+    # subscribers.
+    logger.debug(
+        "6. waiting for all successfully created Tool Meister processes"
+        " to show up as subscribers"
+    )
+    ret_val = wait_for_tds(to_client_chan, logger)
+    if ret_val != 0:
+        if cli_params.redis_server is None:
+            # We created the Redis server, so we should clean it up.
+            return kill_redis_server(redis_pid, ReturnCode.TDSWAITFAILURE)
         else:
-            ret_val = 0
-    else:
-        logger.warning(
-            "unable to successfully start any tool meisters,"
-            " but encountered no failures either: terminating"
-        )
-        ret_val = kill_redis_server(redis_pid)
+            return ReturnCode.TDSWAITFAILURE
+
+    # Setup a Client API object using our existing to_client_chan object to
+    # drive the following client operations ("sysinfo" [optional] and "init"
+    # [required]).
+    with Client(
+        redis_server=redis_server,
+        channel_prefix=cli_tm_channel_prefix,
+        to_client_chan=to_client_chan,
+        logger=logger,
+    ) as client:
+        if sysinfo:
+            sysinfo_path = benchmark_run_dir / "sysinfo" / "beg"
+            try:
+                sysinfo_path.mkdir(parents=True)
+            except Exception:
+                logger.error(
+                    "Unable to create sysinfo-dump directory base path: {}",
+                    sysinfo_path,
+                )
+                ret_val = ReturnCode.EXCSYSINFODIR
+            else:
+                logger.debug("7. Collecting system information")
+                info_log("Collecting system information")
+                ret_val = client.publish(group, sysinfo_path, "sysinfo", sysinfo)
+                ret_val = (
+                    ReturnCode.SUCCESS if ret_val == 0 else ReturnCode.SYSINFOFAILED
+                )
```

Comment:
Why not return directly here as well?

Reply:
This should not feed into the final status of the command, as collecting system information was allowed to fail without affecting the benchmark test run.